### PR TITLE
Trainer refactor into succinct, isolated flow, presets

### PR DESCRIPTION
**Pull request by jules2689**
## What this does
- In `uri_trainer`, `@article_scrape_patterns` doesn't need to be instantiated - so a memoized scrape patterns for the trainer makes sense
- `raw_html` is referred to as `payload` when I traced through the code - which do we want?
#### URI Trainer
- Contains all the log statements
- `train` iterates through the list of data types and fetches the pattern from a private method
- Will save the resulting patterns
- Pattern was in "presets selector", but that class didn't seem necessary and the method was actually just getting a pattern by calling some presets stuff. So I separated it to have a pattern here and call `preset`
#### Preset
- `preset.select` will guide you through selecting from `options` and return nil if skip, and otherwise return the corresponding `selected_option`
- `options` simply returns the list of possible options e.g.

```
1) og_descriptions: example
2) I will provide a pattern using xpath
3) I will provide a pattern using css
4) skip
```
- `transform_results` is the results from the xpath/css patterns for use in `options`
- Realized that `results`, which was originally used in this if/else

```
 if preset_results.empty?
            CLI.log("No presets were found for #{target_data_type}. Skipping to next.")
```

is actually not needed because it really just checks if `data_type_presets` is there and not empty... so now its private

### CODE REVIEW

**Comment by jules2689** [comment]
@richardwu

File: lib/news_scraper/trainer/uri_trainer.rb [hunk 1]
```diff
@@ -1,57 +1,74 @@
+require 'pry'
```
Comment (richardwu) [inline lib/news_scraper/trainer/uri_trainer.rb]: remove?

Comment (jules2689) [reply lib/news_scraper/trainer/uri_trainer.rb]: ya, ignore the little things like this

File: lib/news_scraper/trainer/uri_trainer.rb [hunk 2]
```diff
@@ -1,57 +1,74 @@
...
 
-          CLI.put_header('Save preset')
-          CLI.log_lines(selected_presets.to_yaml)
-          if CLI.confirm("Save these scrape patterns for #{@root_domain}?")
-            save_selected_presets(selected_presets)
-          end
+      def pattern(data_type)
+        CLI.put_header("Determining information for #{data_type}")
+        data_type_presets = article_scrape_patterns['presets'][data_type]
+        pattern = if data_type_presets.nil? || data_type_presets.empty?
```
Comment (richardwu) [inline lib/news_scraper/trainer/uri_trainer.rb]: Checking for `empty?` is too defensive IMO. We'll never add a `data_type` to the YAML file without a preset.

Comment (jules2689) [reply lib/news_scraper/trainer/uri_trainer.rb]: wasnt too sure, it was there before. but ya you're right

File: lib/news_scraper/trainer/preset.rb [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+      end
+
+      def transform_results
+        return @results if @results
+        return {} unless @data_type_presets
+
+        scrape_details = blank_scrape_details
+        @results = @data_type_presets.each_with_object({}) do |(preset_name, preset_details), hash|
+          scrape_details[@data_type] = preset_details
+          train_transformer = Transformers::Article.new(
```
Comment (richardwu) [inline lib/news_scraper/trainer/preset.rb]: Do this in another PR, but separate the `Transformers::Article` for `Trainer` and `Scraper`. The `Scraper` one will only take `uri` and `payload` in it's initializer.

File: lib/news_scraper/trainer/preset.rb [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+            uri: @uri,
+            payload: @payload,
+            scrape_details: scrape_details,
+            scrape_patterns: @scrape_patterns
```
Comment (richardwu) [inline lib/news_scraper/trainer/preset.rb]: You don't need to pass this in since it's only used for getting the `data_type`s, and it defaults to the cached one.

File: lib/news_scraper/trainer/preset.rb [hunk 5]
```diff
@@ -0,0 +1,76 @@
+module NewsScraper
+  module Trainer
+    class Preset
+      PROVIDER_PHRASE = 'I will provide a pattern using'
+
+      def initialize(uri:, scrape_patterns:, payload:, data_type_presets:, data_type:)
```
Comment (richardwu) [inline lib/news_scraper/trainer/preset.rb]: You can remove `scrape_patterns` since you don't need to pass it into `Transformers::Article` (see below comment).

File: lib/news_scraper/trainer/preset.rb [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+        selected_index = options[selected_option]
+        selected_preset_code = transform_results[selected_index].first
+        @data_type_presets[selected_preset_code]
+      end
+
+      private
+
+      def options
+        return @options if @options
```
Comment (richardwu) [inline lib/news_scraper/trainer/preset.rb]: nit but you can do:

```
@options ||= begin
   temp_options = transform_results....
   ...
   temp_options.merge('skip' => 'skip')
end
```

File: lib/news_scraper/trainer/uri_trainer.rb [hunk 7]
```diff
@@ -1,57 +1,74 @@
...
-          CLI.put_footer
-          selected_presets = Trainer::PresetsSelector.new(uri: @uri).select(@raw_html)
+      private
 
-          CLI.put_header('Save preset')
-          CLI.log_lines(selected_presets.to_yaml)
-          if CLI.confirm("Save these scrape patterns for #{@root_domain}?")
-            save_selected_presets(selected_presets)
-          end
+      def pattern(data_type)
```
Comment (richardwu) [inline lib/news_scraper/trainer/uri_trainer.rb]: `selected_pattern`?

File: lib/news_scraper/trainer/uri_trainer.rb [hunk 8]
```diff
@@ -1,57 +1,74 @@
...
+        @payload = Extractors::Article.new(uri: @uri).extract
       end
 
       def train
-        unless @article_scrape_patterns['domains'].key?(@root_domain)
-          CLI.put_header
-          CLI.log("There is no scrape pattern defined for #{@root_domain}"\
-            " in config/article_scrape_patterns.yml")
-          no_scrape_defined
+        return if article_scrape_patterns['domains'].key?(@root_domain)
```
Comment (richardwu) [inline lib/news_scraper/trainer/uri_trainer.rb]: 👌

File: lib/news_scraper/trainer/preset.rb [hunk 9]
```diff
@@ -0,0 +1,76 @@
+module NewsScraper
+  module Trainer
+    class Preset
```
Comment (richardwu) [inline lib/news_scraper/trainer/preset.rb]: Nit: I would rather this be called `PresetX` (I think `PresetSelector` is fine) instead of `Preset` since it implies that it's a class for presets, which it's not really.

File: lib/news_scraper/trainer/preset.rb (hunk 5 above)
Comment (richardwu) [inline lib/news_scraper/trainer/preset.rb]: Nit: Maybe re-order the params so it's

```
data_type:, data_type_presets:, uri:, payload:
```

I feel like `data_type` stuff are the "primary" params. What do you think?

File: lib/news_scraper/trainer/preset.rb [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+        return if selected_option == 'skip'
+
+        selected_index = options[selected_option]
+        selected_preset_code = transform_results[selected_index].first
+        @data_type_presets[selected_preset_code]
+      end
+
+      private
+
+      def options
```
Comment (richardwu) [inline lib/news_scraper/trainer/preset.rb]: Nit: I feel like this method name is too vague.

I get that this is all under the `Preset` class, but I feel like calling it `pattern_options` or `preset_options` would be more descriptive.

File: lib/news_scraper/trainer/preset.rb (hunk 9 above)
Comment (jules2689) [reply lib/news_scraper/trainer/preset.rb]: `PresetSelector.select` just seems weird, but eh

File: lib/news_scraper/trainer/preset.rb [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+        @options['skip'] = 'skip'
+        @options
+      end
+
+      def transform_results
+        return @results if @results
+        return {} unless @data_type_presets
+
+        scrape_details = blank_scrape_details
+        @results = @data_type_presets.each_with_object({}) do |(preset_name, preset_details), hash|
```
Comment (richardwu) [inline lib/news_scraper/trainer/preset.rb]: Nit: same thing here

```
return {} unless @data_type_presets

@results ||= @data_type_presets.each_with_object...
   scrape_details = blank_scrape_details.merge(@data_type => preset_details)
   ...
end
```

then you can memoize `blank_scrape_details`.

**Comment by richardwu** [comment]
Besides a few nits, LGTM 👍 

You could probably reduce the # of tests down with this refactor and have more end-to-end tests.

File: lib/news_scraper/trainer/preset.rb (hunk 11 above)
Comment (jules2689) [reply lib/news_scraper/trainer/preset.rb]: > then you can memoize blank_scrape_details.

already had that done locally :)

File: lib/news_scraper/trainer/uri_trainer.rb [hunk 12]
```diff
@@ -2,55 +2,56 @@ module NewsScraper
...
+        CLI.log "Fetching information..."
+        CLI.put_footer
+
+        selected_presets = {}
+        article_scrape_patterns['data_types'].each do |data_type|
+          selected_presets[data_type] = selected_pattern(data_type)
         end
+
+        save_selected_presets(selected_presets)
+        selected_presets
```
Comment (richardwu) [inline lib/news_scraper/trainer/uri_trainer.rb:23]: Why return this?

Comment (jules2689) [reply lib/news_scraper/trainer/uri_trainer.rb:23]: woops, dont need this. Pushed a minor cleanup to master

**Comment by richardwu** [comment]
Tests are beautiful 👌
